cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+* Positive infinity is of type `FLOAT`, equal to itself and larger than any other number (excluding `NaN`).
+* Negative infinity is of type `FLOAT`, equal to itself and smaller than any other number (excluding `NaN`).
+* Comparing a number to a `NaN` always yields `null`.
+* Booleans are compared such that `false < true`.
+* Strings are compared in dictionary order, i.e. characters are compared pairwise in order and characters missing in a shorter string are considered to be smaller than any other character. For example, `'a' < 'aa'`.
+* Lists are compared in dictionary order, i.e. list elements are compared pairwise in order and elements missing in a shorter list are considered to be smaller than any other value. For example, `[1] < [1,0]`.
+* Lists containing `null` values are incomparable, i.e. comparing them with any other value yields `null`.
+* Maps are compared as if they where lists of lists, where the inner lists are the key-value pairs of the maps, ordered in ascending sort order for the keys. Comparing a map to a value that is not a map always yields `null`.
+* Maps containing explicit `null` value entries are incomparable, i.e. comparing them with any other value yields `null`.
+* The comparison order for nodes (and relationships resp.) is undefined, and an implementation may choose an order for these freely, e.g. by using some kind of id. This order must not change during the execution of the query (i.e. it must be stable in the presence of updates). Comparing a node to a relationship always yields `null`.
+* Paths are compared as if they where a list of alternating nodes and relationships of the path from the start node to the end node. Comparing a path to a value that is not a path always yields `null`.
+* Implementations may choose to define suitable comparability rules for values of additional, non-canonical types. However comparing a value of an additional, non-canonical type to any of the canonical types always yields `null`.
+* Comparing any kind of value not covered above to an arbitrary other value, always yields `null`. This includes comparing a non-`null` value to a `null` value.
+
+=== Orderability
+
+We further propose that orderability be defined between any pair of values, where the result is always `true` or `false`; i.e. always defined.
+To accomplish this, there must be a pre-determined type order and each value must fall under exactly one type in this order. We propose to use the following ascending global sort order of types:
+
+* `NODE`
+* `RELATIONSHIP`
+* `PATH`
+* `MAP`
+* `LIST OF ANY?`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER` (`NaN` is treated as the largest number in orderability only, i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
```

Comment:
The type order is a very important concept, albeit not the simplest (it is very subtle!), and in light of this, I think that an example here would be a very good thing. Something along the lines of if two values, v1 and v2 were ordered, and v1 was a NODE whereas v2 was a MAP, v1 would be ordered before v2? Otherwise, if a misunderstanding occurs here, the reader is lost...